### ISSUE TITLE
Option to trigger notifications based on severity of the CVE

### ISSUE THREAD

**Issue by tapmch** [DependencyTrack/dependency-track#3767]
### Current Behavior

I can configure an alert notification for newly detected vulnerabilities.

### Proposed Behavior

Currently, it is possible to configure a notification when a new vulnerability is detected. I would like to be able to trigger an alert notification only when the severity level is critical or high. 

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this enhancement was already requested

**Comment by farsheedify** [comment]
This feature would be incredibly helpful in reducing the number of notifications sent to our security champions. Currently, some teams in our organization manage numerous microservices, and they receive notifications about new vulnerabilities for every severity level. This results in a flood of emails, which can lead to important messages being ignored due to the sheer volume.

By filtering notifications based on severity, we could limit the emails to only high and critical vulnerabilities. This would significantly reduce the number of emails, ensuring that only the most important notifications are sent, making it easier for our security champions to stay focused on critical issues.

**Comment by Hunroll** [comment]
We also need this feature.
Actually our usecase would be to separate only Critical vulnerabilities and give them e.g. Warning level.

If you have a vision, how it should be configured - I could take this task.

Right now I see it as new setup page (Notifications -> Severity Mapping) which will have such a grid:

| Severity | Notification level |
|----------|--------------------|
| Critical | ERROR |
| High | WARNING |
| Medium | INFORMATIONAL |
| Low | INFORMATIONAL |
| Unassigned | INFORMATIONAL |

I'm not sure if it's OK to tag @nscuro , sorry if it's not. But I'm ready to help with this one.

Thanks!

**Comment by emyhrberg** [comment]
PR's posted @nscuro

**Comment by Jamie-de-Leest** [comment]
Im currently facing the same issue, and i was wondering if there is still someone working on getting this released?

**Comment by iliescuioana** [comment]
+1

**Comment by mattdibi** [comment]
+1

**Comment by lucapseco** [comment]
+1

**Comment by whelks-chance** [comment]
+1, thanks. We have a Slack channel full of automated messages and it's a touch noisy.

**Comment by dmytrobabeshko** [comment]
+1

**Comment by nscuro** [comment]
May I kindly ask everyone to please just upvote the issue instead of posting "+1"? See also: https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#looking-for-existing-issues

As for the issue itself, this is clearly highly requested, but I'd like the requirements to be fleshed out more.

Assuming we add checkboxes for severities as done in #4879, what do you expect to happen?

1. `NEW_VULNERABILITY` seems obvious: Just don't send notifications where the severity doesn't match any of the configured ones.
2. What about `NEW_VULNERABLE_DEPENDENCY`? It contains potentially multiple vulnerabilities. 
   1. Do you expect the list of vulnerabilities to be *filtered* by severity, e.g.:
      1. omit those that don't match the severities, and
      2. drop notification entirely when no vulns left?
   2. Or should we still include all vulns in case *at least one* matches the configured severities?

I personally believe *ii* to be the better option for point *2*, but curious what people *expect*.

Also, do we expect similar requirements for EPSS, CVSS scores, or potentially KEV? I'd like to avoid bolting on more and more filtering options that will be hard to orchestrate and reason about, so if we anticipate more requirements like this we should build with them in mind from the start.

**Comment by Hunroll** [comment]
Thank you for the reply!
As for my use case (we have ~60 project-versions tagged as live/RC, ~1000 components in each, ~100 low-medium vulns) I would like if our SOC/PMs recieve notifications only with a list of fresh-found criticals. 
We have a pipeline that blocks release if a fresh build has criticals, but there's no "clean" way to be notified if they appear after the release.

The list of all vulnerabilities would be quite big to process manually, unfortunately.

Thanks!

**Comment by nscuro** [comment]
@Hunroll So you are only using / interested in the `NEW_VULNERABILITY` notification, am I understanding that right?

**Comment by Hunroll** [comment]
> [@Hunroll](https://github.com/Hunroll) So you are only using / interested in the `NEW_VULNERABILITY` notification, am I understanding that right?

Right. I'm specifically interested only in this one type of notifications.
Thanks!

**Comment by schoofseggl** [comment]
The same filter would be extremely useful in the scheduled new_violation_summary email.

**Comment by nscuro** [comment]
@schoofseggl Please elaborate on the exact behaviour you're expecting there. Since that notification is about policy violations, it already is different from what others so far have requested.

**Comment by schoofseggl** [comment]
Sorry if I was unclear: We do have scheduled email notifications for policy violations and vulnerabilities. 

**Vulnerability scheduled email** 

EVERY new finding is reported in this email, not just new critical ones (which we are most interested in). Today we generate a lot of noise to our developers

**Policy scheduled email** 

Similar and connected to it: we have a Policy "warning" for high vulns and a "fail" policy for critical rated ones. We use the schedule notification to generate the summary email - but we cannot configure that we want to be notified about critical violations only. 

This might be a slightly different use case - but the underlying wish to get notified about critical issues mainly is the same (and this is why I mentioned it here).

**Comment by nscuro** [comment]
@schoofseggl I think both of your use cases can already be addressed by modifying the notification templates. You can simply choose to not render vulnerabilities that don't fit your requirements. Here's the relevant part in the email template:

https://github.com/DependencyTrack/dependency-track/blob/e7523d8b573811a0e350072ee6249a9704ce437b/src/main/resources/templates/notification/publisher/email.peb#L163-L210

The vulnerabilities passed to the template are already grouped by severity.

Here's how you can configure custom templates: https://docs.dependencytrack.org/integrations/notifications/#creation-of-publisher

I agree that it seems related, but I'd like to work out if we're looking at a simple feature that decides whether a notification is sent *at all*, or a much more complex one that needs to filter and modify notification payloads on-the-fly.

**Comment by nscuro** [comment]
Quickly threw this together: https://github.com/DependencyTrack/hyades-apiserver/pull/1924

The idea being that you get to specify a [CEL](https://cel.dev/) expression. When an expression is specified, the notification will only be sent when it matches the expression.

To only emit `NEW_VULNERABILITY` notifications when the vulnerability's severity is `CRITICAL` for example, you'd write:

```js
subject.vulnerability.severity == 'CRITICAL'
```

Or for `NEW_VULNERABLE_DEPENDENCY`, only send when *at least one* vulnerability is `CRITICAL`:

```js
subject.vulnerabilities.exists(vuln, vuln.severity == 'CRITICAL')
```

Plus, you can access *all* fields of the notification, so this would also work for CVSS scores, vectors, potentially EPSS or any other field we may add in the future.

It'd also be easier for us to support, as we don't have to add new code for every new filtering use case.

Any thoughts on this?

**Comment by Hunroll** [comment]
That will totally cover my needs. We already use custom templates, so this CEL filtering would work just fine!
Great idea 👍

**Comment by nscuro** [comment]
@Whatsonyourmind No advertisement of commercial AI slop here please.

**Comment by nscuro** [comment]
Will ship with v5: https://dependencytrack.github.io/hyades/snapshot/usage/notifications/filter-expressions/

**Comment by cpriyakant** [comment]
Can we set an alert based on projects? instead for entire portfolio, This will help in sending alerts to different recipient based on their projects/product.

**Comment by Hunroll** [comment]
> Can we set an alert based on projects? instead for entire portfolio, This will help in sending alerts to different recipient based on their projects/product.

Sure! You can do it even now. In Alerts there's a "Limit to" dropdown which will expand 2 extra fields - projects and tags. Also you can select a parent "aggregator" project and alerts will work for all its children.

<img width="1380" height="439" alt="Image" src="https://github.com/user-attachments/assets/29774767-c388-496a-a0e6-b9a7be6ca808" />

**Comment by cpriyakant** [comment]
@Hunroll  Thanks its working, I don't know how i missed that.

Glad to know it supports parent project because I guess Parent Project doesn't work when assigning parent project to a user or team,   it doesn't aggregate child project using portfolio access control.
